### PR TITLE
CP-50955 Add new-style NBD servers

### PR DESCRIPTION
**Pull request by TimSmithCtx**

### CODE REVIEW

File: drivers/tapdisk-nbdserver.c [hunk 1]
```diff
@@ -440,250 +539,168 @@ receive_newstyle_options(td_nbdserver_client_t *client, int new_fd, bool no_zero
...
+				if(receive_info(new_fd, (char *)buf, opt_len) < 0) {
+					ERR ("Failed to received data for NBD_OPT_EXPORT_NAME");
+					free(buf);
+					goto fail;
+				}
+				buf[opt_len] = '\0';
+				INFO("Exportname %s", buf);
+
+				bzero(&handshake_finish, sizeof handshake_finish);
+				handshake_finish.exportsize = htobe64(NBD_EXPORTSIZE(server));
```
Comment (MarkSymsCtx) [inline drivers/tapdisk-nbdserver.c:577]: Some inconsistent tabbing here?

**Review by MarkSymsCtx — COMMENTED** [review]
Generally looks OK but possibly some inconsistent tab indentation.

File: drivers/tapdisk-nbdserver.c (hunk 1 above)
Comment (TimSmithCtx) [reply drivers/tapdisk-nbdserver.c:577]: Ah yes. There was a mix of tabs and spaces around there originally, which confused matters. Fixed

Comment (MarkSymsCtx) [reply drivers/tapdisk-nbdserver.c:577]: In some copious free time I need to run uncrustify over all this code once and for all.

**Review by MarkSymsCtx — APPROVED** [review]

**Review by KevinLCtx — DISMISSED** [review]

File: drivers/tapdisk-nbdserver.c [hunk 2]
```diff
@@ -427,6 +448,84 @@ send_meta_context (int new_fd, uint32_t reply, uint32_t context_id, const char *
...
+		goto fail;
+	exportnamelen = be32toh(exportnamelen);
+
+	if (exportnamelen > MAX_NBD_EXPORT_NAME_LEN) {
+		ERR("Received %u as export name length which exceeds the maximum of %u",
+			exportnamelen, MAX_NBD_EXPORT_NAME_LEN);
+		goto fail;
+	}
+
+	exportname = malloc(exportnamelen + 1);
```
Comment (rdn32) [inline drivers/tapdisk-nbdserver.c:479]: exportname only appears to be freed in error case.

Comment (MarkSymsCtx) [reply drivers/tapdisk-nbdserver.c:479]: Well spotted.

**Review by MarkSymsCtx — CHANGES_REQUESTED** [review]
Fix the memory leak

**Comment by TimSmithCtx** [comment]
> Fix the memory leak

There was also a mysterious reversion of NBD_OPT_MAGIC to NBD_NEW_VERSION which I've corrected.

**Review by MarkSymsCtx — APPROVED** [review]

**Review by rdn32 — APPROVED** [review]

**Pull request merged**